Sources/NIOSSH/Keys And Signatures/NIOSSHPublicKey.swift

```diff
@@ -175,16 +188,107 @@ extension NIOSSHPublicKey {
             return Self.ecdsaP384PublicKeyPrefix
         case .ecdsaP521:
             return Self.ecdsaP521PublicKeyPrefix
+        case .custom(let publicKey):
+            return publicKey.publicKeyPrefix.utf8
         case .certified(let base):
             return base.keyPrefix
         }
     }
 
+    private static let bundledAlgorithms: [String.UTF8View] = [
+        Self.ed25519PublicKeyPrefix, Self.ecdsaP384PublicKeyPrefix, Self.ecdsaP256PublicKeyPrefix, Self.ecdsaP521PublicKeyPrefix,
+    ]
+
     internal static var knownAlgorithms: [String.UTF8View] {
-        [Self.ed25519PublicKeyPrefix, Self.ecdsaP384PublicKeyPrefix, Self.ecdsaP256PublicKeyPrefix, Self.ecdsaP521PublicKeyPrefix]
+        bundledAlgorithms + customPublicKeyAlgorithms.map { $0.publicKeyPrefix.utf8 }
+    }
+
+    internal static var customPublicKeyAlgorithms: [NIOSSHPublicKeyProtocol.Type] {
+        _CustomAlgorithms.publicKeyAlgorithmsLock.withLock {
+            _CustomAlgorithms.publicKeyAlgorithms
+        }
+    }
+
+    internal static var customSignatures: [NIOSSHSignatureProtocol.Type] {
+        _CustomAlgorithms.signaturesLock.withLock {
+            _CustomAlgorithms.signatures
+        }
+    }
+}
+
+public enum NIOSSHAlgorithms {
```

Comment:
I remain uncertain about doing this with global state. Can we consider placing this on per connection configuration instead of using this global config?

Reply:
@Joannis Can you weigh in on this?

Reply:
I agree that a per-connection basis is a better solution. But we will need a registry for sure, since we can have a variety of Host Keys for clients, and accepted public keys for servers

Reply:
@Lukasa @gwynne While I was picking this up I was discussing this with Joannis and want to introduce a per connection solution. But while discussing with the way the custom keys currently work we could get rid of the BackingKey enums that are currently internal in the library. This would eliminate a lot of switch statements and simplify a lot of methods quite a bit. We don't immediately see any issues, but I wanted to check in with you, what are your thoughts on this?

Reply:
I don't think the switch statements are that big a deal: they provide fast-paths for the most common use-cases where the compiler can be aware of exactly what the implementation is. The extensible backing pattern should certainly run this through a protocol though.

Reply:
That's true it would make the whole implementation more consistent however and remove a lot of boilerplate/duplication. So not something the user benefits from perse but still could be interesting in terms of maintainability/accessible for new devs checking out the repo? I put in quite some work to get it towards that end. Would you be interested in seeing a PR? (it still needs a little work but if you're not interested in merging that kind of work in to the repo I'll probably leave it as is and just finish up your last remarks regarding not using global state for registering algorithms, which btw already needs quite a bit of moving around of methods in order to pass the registered algorithms down to the right ByteBuffer extensions) I'm hoping to finish up the work and then ideally divide it into a few PR's so it's a bit more manageable for you to review. Anyway let me know! :)